modules/launcher/src/main/java/org/mule/module/launcher/artifact/AbstractArtifactClassLoader.java

```diff
@@ -60,9 +69,62 @@ public void dispose()
                 logger.error(e);
             }
         }
+
+        LeakCleaner leakCleaner = getLeakCleanerInstance();
+        if (leakCleaner != null)
+        {
+            leakCleaner.clean();
+        }
+
         super.dispose();
     }
 
+    public String getLeakCleanerClassName()
+    {
+        return StringUtils.defaultString(leakCleanerClassName, DEFAULT_LEAK_CLEANER_CLASS);
+    }
+
+    public void setLeakCleanerClassName(String leakCleanerClass)
+    {
+        this.leakCleanerClassName = leakCleanerClass;
+        leakCleaner = null;
+    }
+
+    public LeakCleaner getLeakCleanerInstance()
+    {
+        if(leakCleaner==null)
+        {
+            leakCleaner = createLeakCleanerInstance(getLeakCleanerClassName());
+        }
+        return leakCleaner;
+    }
+
+    private LeakCleaner createLeakCleanerInstance(String className)
+    {
+        InputStream classStream = this.getClass().getResourceAsStream(className);
+        if (classStream == null)
+        {
+            classStream = AbstractArtifactClassLoader.class.getResourceAsStream(className);
+        }
+        if (classStream == null)
+        {
+            logger.warn("Could not find leak cleaner class with name: " + className);
```

Comment:
Could be, but if we raise an error the entire dispose mechanism will fail, and here we want to improve the cleaning. If we raise an error the leak will continue and other resources could remains in memory too.